source/common/config/subscription_factory.h

```diff
@@ -79,14 +91,33 @@ class SubscriptionFactory {
       break;
     }
     case envoy::api::v2::core::ConfigSource::kAds: {
-      result.reset(new GrpcMuxSubscriptionImpl<ResourceType>(cm.adsMux(), stats));
+      result.reset(new GrpcMuxSubscriptionImpl<ResourceType>(ads_mux, stats));
       break;
     }
     default:
       throw EnvoyException("Missing config source specifier in envoy::api::v2::core::ConfigSource");
     }
     return result;
   }
+
+  static std::unique_ptr<Subscription<envoy::api::v2::Cluster>> cdsSubscriptionFromConfigSource(
+      const envoy::api::v2::core::ConfigSource& cds_config, const Optional<envoy::api::v2::core::ConfigSource>& eds_config,
+      const LocalInfo::LocalInfo& local_info,
+      Event::Dispatcher& dispatcher, Upstream::ClusterManager& cm, Runtime::RandomGenerator& random,
+      Stats::Scope& scope, Config::GrpcMux& ads_mux) {
+
+    return subscriptionFromConfigSource<envoy::api::v2::Cluster>(
+          cds_config, local_info.node(), dispatcher, cm, random, scope,
+          [&cds_config, &eds_config, &cm, &dispatcher, &random, &scope,
+           &local_info]() -> Config::Subscription<envoy::api::v2::Cluster>* {
+            return new Upstream::CdsSubscription(Config::Utility::generateStats(scope), cds_config,
+                                       eds_config, cm, dispatcher, random, local_info);
+          },
+          ads_mux,
+          "envoy.api.v2.ClusterDiscoveryService.FetchClusters",
```

Comment:
See discussion in https://github.com/envoyproxy/data-plane-api/pull/524, I wonder if lazy deserves a new endpoint or not.